Add Ollama as a supported provider for plugin-knowledge and embeddings

david-dina opened the pull request:
Added full support and compatibility for Ollama via plugin-ollama. Fully tested on Windows 10.

0xbbjoker commented:
<img width="745" height="485" alt="image" src="https://github.com/user-attachments/assets/003530db-1ff9-412e-930c-6d2fd15e574c" />

It doesn't work on my side. I am on macos. 

Also I don't think this does make sense. Embedding part is not necessary at all. Should use default: `runtime.useModel(TEXT_EMBEDDING, text)`.

@david-dina  Which models are you running with ollama btw?

david-dina commented:
I didnt think of using runtime so ill refactor those. As for Models I have a pretty beefy setup so those are: 
OLLAMA_SMALL_MODEL=mistral-nemo-8k
OLLAMA_MEDIUM_MODEL=gemma3-8k
OLLAMA_LARGE_MODEL=command-r
OLLAMA_EMBEDDING_MODEL=bge-m3

david-dina commented:
I just tested it with MacOS and using my original setup as my Ollama URL. worked as expected. 
<img width="1122" height="225" alt="image" src="https://github.com/user-attachments/assets/06df7c3d-af1a-45f1-9faa-8a6be9d55f51" />
I ran the plugin via `elizaos start` as normal did u run yours differently, or get anymore detailed errors? This may also be a issue on the Model Generation side
@0xbbjoker

david-dina commented:
regarding the use TEXT_EMBEDDING this plugin uses OLLAMA_EMBEDDING_MODEL strictly from plugin-ollama and doesnt create the use of it itself. Are you saying I should remove the dependancy??? as to use the runtime text embedding? or im just a little confused as to what your looking for.

0xbbjoker commented:
@david-dina you can see here https://github.com/elizaos-plugins/plugin-knowledge/pull/34#issuecomment-3114405176

Yours is also failing to generate a chunk.

Notice: `1 chunks failed processing` 

Also I see you have an error. 

What I wanna say for embedding: There is no need to do it manually and you have to use `plugin-ollama` for embeddings generally for the agent if you want to use it for the `knowledge` as well. So no need to add this part. 

<img width="361" height="189" alt="image" src="https://github.com/user-attachments/assets/5c87fd9f-e6d8-4aff-aba3-dae9e0f27f6d" />

If you setup plugins like this, it should work out of the box for embeddings.

david-dina commented:
@0xbbjoker, It should work now with this https://github.com/elizaos-plugins/plugin-ollama/pull/11 Worked on mine, with chunking. Also removed custom configs as Ollama does actually export to Text_Large/Small/Embedding already, it just wasnt working for me at the time. Then I added custom configs for Rate limiting.

0xbbjoker reviewed (changes requested):
@david-dina there are some fundamental issues with this approach:

## 1. Embeddings Already Work

The embedding implementation for Ollama is unnecessary. When using `plugin-ollama`, embeddings work out of the box through the runtime. The plugin already provides `OLLAMA_EMBEDDING_MODEL` (defaults to `nomic-embed-text`) without requiring any custom provider code.

## 2. Model Detection

The key issue is detecting which Ollama model is being used. The `plugin-ollama` registers `OLLAMA_LARGE_MODEL` as `TEXT_LARGE` in the runtime. For contextual retrieval to work, we need to:
- Read the `OLLAMA_LARGE_MODEL` setting from the runtime
- Verify it supports 32k+ context window
- Only then enable contextual retrieval

## 3. Context Window Requirements

Contextual retrieval is designed for models with **128k+ context windows** because it's typically used for:
- Large datasets (analytics, research papers)
- Full books or lengthy documents
- Complex documents requiring smart chunking with full context

With smaller models (32k or less), contextual retrieval provides minimal benefit and often causes more problems than it solves.

## Critical questions:

- Which specific Ollama models are you testing with (`OLLAMA_LARGE_MODEL` value)?
- Have you tested with documents larger than 4k tokens?
- **What's your use case for contextual retrieval with local models?** This feature is typically used for processing large documents that require the full context for intelligent chunking.
- How are you handling the concurrent request limitations of local models?

## The implementation should:

- Use the existing runtime embedding support (remove custom embedding code)
- Detect the model's context window size before enabling contextual retrieval
- Only enable for models with 128k+ context (like `llama3.1:8b` or `qwen2.5:7b`)
- Warn users if their model doesn't meet requirements

I am still having issues on macos with small model which just can't handle this.

0xbbjoker commented:
Not necessary as embeddings got fixed in the `plugin-ollama`. Closing!
